skimage/morphology/_util.py

```diff
@@ -51,35 +54,206 @@ def _validate_connectivity(image_dim, connectivity, offset):
     return c_connectivity, offset
 
 
-def _offsets_to_raveled_neighbors(image_shape, structure, center, order='C'):
+def _offsets_to_raveled_neighbors(image_shape, selem, center, order='C'):
     """Compute offsets to a samples neighbors if the image would be raveled.
 
     Parameters
     ----------
     image_shape : tuple
         The shape of the image for which the offsets are computed.
-    structure : ndarray
+    selem : ndarray
         A structuring element determining the neighborhood expressed as an
         n-D array of 1's and 0's.
-    center : sequence
-        Tuple of indices specifying the center of `selem`.
+    center : tuple
+        Tuple of indices to the center of `selem`.
+    order : {"C", "F"}, optional
+        Whether the image described by `image_shape` is in row-major (C-style)
+        or column-major (Fortran-style) order.
 
     Returns
     -------
-    offsets : ndarray
+    raveled_offsets : ndarray
         Linear offsets to a samples neighbors in the raveled image, sorted by
-        their Euclidean distance from the center.
+        their distance from the center.
+
+    Notes
+    -----
+    This function will return values even if `image_shape` contains a dimension
+    length that is smaller than `selem`.
 
     Examples
     --------
     >>> _offsets_to_raveled_neighbors((4, 5), np.ones((4, 3)), (1, 1))
     array([-5, -1,  1,  5, -6, -4,  4,  6, 10,  9, 11])
+    >>> _offsets_to_raveled_neighbors((2, 3, 2), np.ones((3, 3, 3)), (1, 1, 1))
+    array([ 2, -6,  1, -1,  6, -2,  3,  8, -3, -4,  7, -5, -7, -8,  5,  4, -9,
+            9])
+    """
+    if not selem.ndim == len(image_shape) == len(center):
+        raise ValueError(
+            "number of dimensions in image_shape, selem and center "
+            "does not match"
+        )
+
+    selem_indices = np.array(np.nonzero(selem)).T
+    offsets = selem_indices - center
+
+    if order == 'F':
+        offsets = offsets[:, ::-1]
+        image_shape = image_shape[::-1]
+    elif not order == 'C':
+        raise ValueError("order was not 'C' or 'F'")
+
+    # Scale offsets in each dimension and sum
+    ravel_factors = image_shape[1:] + (1,)
+    ravel_factors = np.cumprod(ravel_factors[::-1])[::-1]
+    raveled_offsets = (offsets * ravel_factors).sum(axis=1)
+
+    # Sort by distance
+    distances = np.abs(offsets).sum(axis=1)
+    raveled_offsets = raveled_offsets[np.argsort(distances)]
```

Comment:
I haven't been able to figure out a way to keep the same order for neighbors of the same connectivity. Maybe that's due to `argsort` defaulting to the unstable `quicksort`? This does affect in which order algorithms iterate over neighbors with the same distance. I don't think this relevant for the algorithms relying on this function though and the test suite seems to back this up...

Reply:
Do I understand correctly that the problem is that the order of indices is not always the same? (lack of reproducibility)

If yes, why not add another information to disambiguate the distance, such as a raveled index within the structuring element? You could then use the `order` parameter of `argsort`, see for example
```
>>> a = np.array([(2, 10), (2, 5), (1, 3), (1, 2), (3, 1)], dtype=[('distance', 'f8'), ('index', 'f8')])                                                
>>> np.sort(a, order=['distance', 'index'])                                                                                                             
array([(1.,  2.), (1.,  3.), (2.,  5.), (2., 10.), (3.,  1.)],
      dtype=[('distance', '<f8'), ('index', '<f8')])
>>> np.argsort(a, order=['distance', 'index'])                                                                                                          
array([3, 2, 1, 0, 4])
```

Reply:
> Do I understand correctly that the problem is that the order of indices is not always the same? (lack of reproducibility)

No it's always the same and reproducable but changes between the old and new implementation. E.g. old:
```python
>>> _offsets_to_raveled_neighbors((3, 3, 3), np.ones((3, 3, 3)), (1, 1, 1))
array([ -1,   3,  -9,   1,   9,  -3,   4,   2,  12,  -2,  -4,  -6,  10,
        -8, -10, -12,   8,   6,  11, -13,   5,  -5,  -7, -11,   7,  13])
```
and new:
```python
>>> _offsets_to_raveled_neighbors((3, 3, 3), np.ones((3, 3, 3)), (1, 1, 1))
array([  3,  -9,   1,  -1,   9,  -3,   4,   2,  12,  -2,  -4,  -6,  10,
        -8, -10, -12,   8,   6,  11, -13,   5,  -5,  -7, -11,   7,  13])
```

I vaguely remember fiddling around with  the `kind=` argument to `argsort`. I think we want a stable sort that preserves the order of input arguments (which is memory-order) as long as they have the same distance. So `kind="mergesort"` and switch to `kind="stable"` once we hit NumPy 1.15?

Reply:
Yes. LOL Sorry for the long delay in responding!